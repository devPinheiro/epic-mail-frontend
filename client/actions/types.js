export const REGISTER = "REGISTER";
export const GET_ERRORS = "GET_ERRORS";
export const SIGN_UP = "SIGN_UP";
export const LOGIN = "LOGIN";
export const SET_CURRENT_USER = "SET_CURRENT_USER";
export const GET_PROFILE = "GET_PROFILE";
export const LOADING = "LOADING";
export const LOADING_DONE = "LOADING_DONE";
export const PROFILE_NOT_FOUND = "PROFILE_NOT_FOUND";
export const CLEAR_PROFILE = "CLEAR_PROFILE";
export const OPEN_SIDEBAR = "OPEN_SIDEBAR";
export const CLOSE_SIDEBAR = "CLOSE_SIDEBAR";
export const COMPOSE_MAIL = "COMPOSE_MAIL";
export const COMPOSE_DRAFT = "COMPOSE_DRAFT";
export const SUCCESS_MAIL = "SUCCESS_MAIL";
export const INBOX = "INBOX";
