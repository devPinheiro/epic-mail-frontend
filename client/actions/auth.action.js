import axios from "axios";
import jwt_decode from "jwt-decode";
import { GET_ERRORS, SET_CURRENT_USER } from "./types";
import { setAuthToken } from "../utils";
import { getProfile } from "./profile.action";

export const signUp = (data, history) => dispatch => {
  return axios
    .post("https://epic-mail-devp.herokuapp.com/api/v1/auth/signup", data)
    .then(res => {
      // store user token
      const userToken = res.data.data.token;

      setAuthToken(userToken);
      // decode user
      const user = jwt_decode(userToken);
      // set current user
      dispatch(setCurrentUser(user));
      localStorage.setItem("token", userToken);
      history.push("/dashboard");
    })
    .catch(err => {
      const formatError = {};
      formatError["email"] = err.response.data.message;
      dispatch({
        type: GET_ERRORS,
        payload: formatError
      });
    });
};

export const login = (data, history) => dispatch => {
  return axios
    .post("https://epic-mail-devp.herokuapp.com/api/v1/auth/login", data)
    .then(res => {
      // store user token
      const userToken = res.data.data.token;
      setAuthToken(userToken);
      // decode user
      const user = jwt_decode(userToken);
      // set current user
      dispatch(setCurrentUser(user));
      // set user profile
      dispatch(getProfile(userToken));
      localStorage.setItem("token", userToken);
      history.push("/dashboard");
    })
    .catch(err => {
      const formatError = {};
      formatError["email"] = err.response.data.error;
      dispatch({
        type: GET_ERRORS,
        payload: formatError
      });
    });
};

export const setCurrentUser = decoded => {
  return {
    type: SET_CURRENT_USER,
    payload: decoded
  };
};

export const logout = history => dispatch => {
  // remove token
  localStorage.removeItem("token");
  setAuthToken(false);
  dispatch(setCurrentUser({}));
  history.push("/");
};
