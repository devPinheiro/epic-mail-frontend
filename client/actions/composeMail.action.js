import axios from "axios";
import { GET_ERRORS, SUCCESS_MAIL } from "./types";

export const composeMail = (data, history) => dispatch => {
  const tokenId = localStorage.getItem("token");
  axios
    .post("https://epic-mail-devp.herokuapp.com/api/v1/messages", data, {
      method: "get",
      mode: "cors",
      headers: {
        Accept: "application/json",
        "Content-Type": "application/json",
        "x-access-token": tokenId
      }
    })
    .then(res => {
      // get success
      const result = res.data;
      // dispatch a success message
      if (result.status === 201) {
        dispatch(sendSuccessMessage("Mail sent successfully"));
      }
      history.push("/dashboard");
    })
    .catch(err => {
      dispatch({
        type: GET_ERRORS,
        payload: err.response.data.error
      });
    });
};

export const sendSuccessMessage = payload => {
  return {
    type: SUCCESS_MAIL,
    payload
  };
};
