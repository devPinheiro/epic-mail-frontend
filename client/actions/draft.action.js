import axios from "axios";
import { draft_MAIL_SUCCESS, draft_MAIL_START, draft_MAIL_FAIL } from "./types";

export const draft = () => dispatch => {
  dispatch(draftMailStart());
  const tokenId = localStorage.getItem("token");
  return axios
    .get("https://epic-mail-devp.herokuapp.com/api/v1/messages/draft", {
      method: "get",
      mode: "cors",
      headers: {
        Accept: "application/json",
        "Content-Type": "application/json",
        "x-access-token": tokenId
      }
    })
    .then(res => {
      // get success
      const result = res.data;
      // dispatch a success message

      if (result.status === 200) {
        dispatch(draftMailSuccess(result.data));
      }
      history.push("/dashboard");
    })
    .catch(err => {
      if (err.response) {
        dispatch({
          type: draft_MAIL_FAIL,
          payload: err.response.data.error
        });
      }
    });
};

export const draftMailSuccess = payload => {
  return {
    type: draft_MAIL_SUCCESS,
    payload
  };
};

export const draftMailStart = () => {
  return {
    type: draft_MAIL_START
  };
};

export const draftMailFail = payload => {
  return {
    type: draft_MAIL_FAIL,
    payload
  };
};
