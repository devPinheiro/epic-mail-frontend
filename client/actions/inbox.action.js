import axios from "axios";
import { GET_ERRORS, INBOX, LOADING, LOADING_DONE } from "./types";

export const inbox = () => dispatch => {
  dispatch(inboxLoading());
  const tokenId = localStorage.getItem("token");
  return axios
    .get("https://epic-mail-devp.herokuapp.com/api/v1/messages", {
      method: "get",
      mode: "cors",
      headers: {
        Accept: "application/json",
        "Content-Type": "application/json",
        "x-access-token": tokenId
      }
    })
    .then(res => {
      // get success
      const result = res.data;
      // dispatch a success message

      if (result.status === 200) {
        dispatch(sendSuccessMessage(result.data));
        dispatch(inboxLoadingDone());
      }
      history.push("/dashboard");
    })
    .catch(err => {
      dispatch({
        type: GET_ERRORS,
        payload: err.response.data.error
      });
    });
};

export const sendSuccessMessage = payload => {
  return {
    type: INBOX,
    payload
  };
};

export const inboxLoading = () => {
  return {
    type: LOADING
  };
};

export const inboxLoadingDone = () => {
  return {
    type: LOADING_DONE
  };
};
