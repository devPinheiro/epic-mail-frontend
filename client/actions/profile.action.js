import axios from "axios";
import { GET_PROFILE, LOADING, CLEAR_PROFILE, LOADING_DONE } from "./types";

export const getProfile = token => dispatch => {
  dispatch(profileLoading());
  axios
    .get("https://epic-mail-devp.herokuapp.com/api/v1/auth/user", {
      method: "get",
      mode: "cors",
      headers: {
        Accept: "application/json",
        "Content-Type": "application/json",
        "x-access-token": token
      }
    })
    .then(res => {
      dispatch({
        type: GET_PROFILE,
        payload: res.data
      });
      dispatch(profileLoadingDone());
    })
    .catch(() =>
      dispatch({
        type: GET_PROFILE,
        payload: {}
      })
    );
};

export const profileLoading = () => {
  return {
    type: LOADING
  };
};

export const profileLoadingDone = () => {
  return {
    type: LOADING_DONE
  };
};

export const clearProfile = () => {
  return {
    type: CLEAR_PROFILE
  };
};
