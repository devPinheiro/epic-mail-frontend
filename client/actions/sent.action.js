import axios from "axios";
import { SENT_MAIL_SUCCESS, SENT_MAIL_START, SENT_MAIL_FAIL } from "./types";

export const sent = () => dispatch => {
  dispatch(sentMailStart());
  const tokenId = localStorage.getItem("token");
  return axios
    .get("https://epic-mail-devp.herokuapp.com/api/v1/messages/sent", {
      method: "get",
      mode: "cors",
      headers: {
        Accept: "application/json",
        "Content-Type": "application/json",
        "x-access-token": tokenId
      }
    })
    .then(res => {
      // get success
      const result = res.data;
      // dispatch a success message

      if (result.status === 200) {
        dispatch(sentMailSuccess(result.data));
      }
      history.push("/dashboard");
    })
    .catch(err => {
      if (err.response) {
        dispatch({
          type: SENT_MAIL_FAIL,
          payload: err.response.data.error
        });
      }
    });
};

export const sentMailSuccess = payload => {
  return {
    type: SENT_MAIL_SUCCESS,
    payload
  };
};

export const sentMailStart = () => {
  return {
    type: SENT_MAIL_START
  };
};

export const sentMailFail = payload => {
  return {
    type: SENT_MAIL_FAIL,
    payload
  };
};
