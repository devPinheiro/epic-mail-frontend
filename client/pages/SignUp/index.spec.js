import React from "react";
import { shallow, mount, configure } from "enzyme";
import Adapter from "enzyme-adapter-react-16";
import { BrowserRouter } from "react-router-dom";
import mockAxios from "axios";
import thunk from "redux-thunk";
import configureMockStore from "redux-mock-store";
import { signUp } from "../../actions/auth.action";
import { index as SignUpPage } from "./index";
import signupReducer from "../../reducers/authReducer";
import { SIGN_UP, SET_CURRENT_USER } from "../../actions/types";

configure({ adapter: new Adapter() });

describe("Signup component Tests", () => {
  const defaultProps = {
    signUp: jest.fn(),
    errors: {},
    auth: {}
  };

  it("renders the Signup component correctly", () => {
    const wrapper = shallow(<SignUpPage {...defaultProps} />);
    expect(wrapper).toMatchSnapshot();
  });

  it("should render component successfully and check form interactions", () => {
    const component = mount(
      <BrowserRouter>
        <SignUpPage {...defaultProps} />
      </BrowserRouter>
    );
    component.find("form").simulate("submit");
  });

  it(`should ensure submit button text on sign up is rendered appropriately`, () => {
    const component = mount(
      <BrowserRouter>
        <SignUpPage {...defaultProps} />
      </BrowserRouter>
    );
    expect(component.find(".signup")).toBeTruthy();
  });

  it("should simulate an onchange event on form input", () => {
    const event = {
      preventDefault() {},
      target: { name: "firstName", value: "sammy" }
    };
    const component = mount(
      <BrowserRouter>
        <SignUpPage {...defaultProps} />
      </BrowserRouter>
    );

    const inputTag = component.find("input").at(0);
    inputTag.simulate("change", event);
  });

  it("should simulate an onchange event on form input", () => {
    const event = {
      preventDefault() {},
      target: { name: "lastName", value: "sammy" }
    };
    const component = mount(
      <BrowserRouter>
        <SignUpPage {...defaultProps} />
      </BrowserRouter>
    );

    const inputTag = component.find("input").at(1);
    inputTag.simulate("change", event);
  });

  it("should simulate an onchange event on form input", () => {
    const event = {
      preventDefault() {},
      target: { name: "email", value: "sammy@email.com" }
    };
    const component = mount(
      <BrowserRouter>
        <SignUpPage {...defaultProps} />
      </BrowserRouter>
    );

    const inputTag = component.find("input").at(2);
    inputTag.simulate("change", event);
  });

  it("should call Sign up request function on submit of the form", () => {
    const wrapper = mount(
      <BrowserRouter>
        <SignUpPage {...defaultProps} />{" "}
      </BrowserRouter>
    );

    wrapper.find("form").simulate("submit", { preventDefault() {} });
  });
});

describe("Sign Up async action tests", () => {
  const userCredentials = {
    email: "someuser@gmail.com",
    password: "randompassword"
  };

  const middleware = [thunk];
  const mockStore = configureMockStore(middleware);
  let store;
  beforeEach(() => {
    store = mockStore({});
    jest.resetAllMocks();
  });
  afterEach(() => {
    store.clearActions();
  });

  it("Should trigger the signup action", () => {
    store = mockStore({});
    const mockData = {
      data: {
        firstName: "some",
        lastName: "user",
        email: "someuser@gmail.com",
        token:
          "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpZCI6MjMsIâ€¦c4OX0.DfJClt28K3qIgJJWp9FMypDO0tKD1NoiNtO9YvEOvsQ"
      }
    };

    mockAxios.post.mockResolvedValue({
      status: 200,
      data: mockData
    });

    const expectedActions = [{ payload: mockData.data }];
    const historyObject = {
      push: jest.fn()
    };

    store.dispatch(signUp(userCredentials, historyObject)).then(() => {
      expect(store.getActions()).toEqual(expectedActions);
    });
  });
});

describe("Signup reducer tests", () => {
  it("Should return default state", () => {
    const newState = signupReducer(undefined, {});
    expect(newState).toEqual({
      isAuthenticated: false,
      user: {}
    });
  });

  it("Should return SIGN_UP state", () => {
    const initialState = {
      isAuthenticated: false,
      user: {}
    };
    const newState = signupReducer(initialState, {
      type: SIGN_UP
    });
    expect(newState).toEqual({
      isAuthenticated: false,
      user: undefined
    });
  });

  it("Should return SET_CURRENT_USER state", () => {
    const initialState = {
      isAuthenticated: false,
      user: {}
    };
    const newState = signupReducer(initialState, {
      type: SET_CURRENT_USER
    });
    expect(newState).toEqual({
      isAuthenticated: false,
      user: undefined
    });
  });
});
