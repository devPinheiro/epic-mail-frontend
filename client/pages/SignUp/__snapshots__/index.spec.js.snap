// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Signup component Tests renders the Signup component correctly 1`] = `ShallowWrapper {}`;
